File: problem23.rb

class Integer
  def is_abundant_num?
    sum = 1
    i = 2
    while i <= Math.sqrt(self)
      if self % i == 0 && i ** 2 == self
        sum += self/i
        break   
      elsif self % i == 0
        sum += self/i + i
      end
  	  i += 1
    end

    if sum > self
  	  return true
    else
      return false
    end
  end
end


def not_abundant_sum
  list = []
  i = 1
  while i < 28124
    if i.is_abundant_num?
      list << i
    end
    
  	i += 1
  end
 
  sums = []
  list.each do |first_num|
    list.each do |second_num|
      if first_num + second_num > 28123
        break
      else
        sums << first_num + second_num
      end
    end
  end
 
  ((1..28123).to_a - sums).reduce(:+)
 
end

p not_abundant_sum

File: problem23.py
import math

def abundant_number(num):
  result = 1
  i = 2
  while (i <= math.sqrt(num)):
    if (num % i == 0):
      if(i ** 2 == num):
        result += num / i
      else:
        result += num / i + i
    i +=1
  if (result > num):
    return True
  else:
    return False

def sum_not_abundant():
  abundant_nums = []
  i = 1
  result = []
  while (i < 28124):
    if (abundant_number(i)):
      abundant_nums.append(i)
    i += 1
  for first_num in abundant_nums:
    for second_num in abundant_nums:
      pair_sum = first_num + second_num
      if (pair_sum > 28123):
        break
      else:
        result.append(pair_sum)
  numbers = range(1,28124)
  differences = list(set(numbers) - set(result))
  return reduce(lambda x,y: x + y, differences)

print sum_not_abundant()
